### DISCUSSION TITLE
Restricting local users to send emails to specific domains

### DISCUSSION THREAD

**Discussion by nv160** [rspamd/rspamd#4407] [General]
Hello everyone, I had been searching for such solution where I can define a user list and specific domains where specific users can only send emails to defined domains / email IDs including internal domains.

I tried with some combined map like below:
```
restricted_local_senders {
  type = "combined";
  prefilter = true;
  regexp = true;
  action = "reject";
  rules {
    user {
      map = "/etc/rspamd/override.d/local_restricted_users.map";
      selector = "user";
    }
    rcpt = {
      map = "/etc/rspamd/override.d/allowed_domains_restricted_users.map";
      selector = "rcpts:domain";
    }
  }
  expression = "user & rcpt"
}
```

local_restricted_users.map contains internal user Email IDs and allowed_domains_restricted_users.map contains domains.

Well as Action is REJECT currently it is working fine. 

Just want to know how can I make it reverse... like instead of reject it only allows sending of emails to specified domains. (Changing action to "accept" opens everything.)

**Comment by fatalbanana** [comment]
~~`user & !rcpt`~~

**Reply by fatalbanana** [reply]
If you want to ensure that **all** recipients match (probably yes you do) you may need to reach for Lua I suppose.

**Reply by topialife** [reply]
Hi can you help out with Lua for matching all recipients? Actually I don't have a single knowledge on it.

**Reply by fatalbanana** [reply]
~~~lua
-- /etc/rspamd/rspamd.local.lua

local lua_maps = require 'lua_maps'

local allowed_domains_map = lua_maps.map_add_from_ucl(
  '/etc/rspamd/allowed_domains_restricted_users.map',
  'set',
  'Domains restricted users should be allowed to email'
)

local restricted_users_map = lua_maps.map_add_from_ucl(
  '/etc/rspamd/restricted_users.map',
  'set',
  'Usernames of restricted users'
)

rspamd_config.USER_RESTRICTION = {
  callback = function(task)
    if not (allowed_domains_map and restricted_users_map) then return end
    local username = task:get_user()
    if not username then return end
    if not restricted_users_map:get_key(username) then return end
    local rcpts = task:get_recipients('smtp')
    if not rcpts then return end
    for _, rcpt in ipairs(rcpts) do
      if not allowed_domains_map:get_key(rcpt.domain) then
        task:set_pre_result({action = 'reject'})
        return true, string.format('%s->%s', username, rcpt.domain)
      end
    end
  end,
}
~~~

**Reply by topialife** [reply]
I am using with mailcow rspamd and rspamd.local.lua file is in rspamd/lua/ directory. Also file is also having other configs. Shall I append this at the end of the file?

**Reply by topialife** [reply]
Tried it adding in /etc/rspamd/lua/rspamd.local.lua file at the end of the file but not working. may be I am missing something.

**Reply by topialife** [reply]
hello... if you can get some time to solve it please.

**Comment by topialife** [comment]
RESTRICTED_LOCAL_DOMAINUSER {
  type = "combined";
  prefilter = true;
  #regexp = true;
  action = "reject";
  message = "You are not allowed to send Email to this Domain Address.";
  rules {
    user {
      map = "/etc/rspamd/override.d/local_delivery_only.map";
      selector = "user";
    }
    rcpt = {
      map = "/etc/rspamd/override.d/local_delivery_allowed_domains.map";
      selector = "rcpts:domain";
    }
  }
  expression = "user & !rcpt"
}


This config is working fine but only when users tries to send to all domains that are restricted... 

but if a user sends an email to a domain which is allowed in list (e.g. local_delivery_allowed_domains.map) and another email ID which is NOT allowed in the list... email goes to all domains including the not allowed ones.
